addons/mail/models/mail_thread.py

```diff
@@ -1807,7 +1808,8 @@ def message_post(self, *,
         record_name = record_name or self.display_name
 
         # Find the message's author
-        if self.env.user._is_public() and 'guest' in self.env.context:
+        # connected users bypass the guest key context to avoid users with a remaining guest cookie on their session
+        if (not request or not request.session.uid) and 'guest' in self.env.context:
```

Comment:
Could someone explain how we end up with that code ?

Reply:
Do you want me to say "because we don't have guest_id param", or are you actually asking why each part of the condition is necessary considering the current flawed approach?


Well here it is anyway:

The main goal is what is explained the comment: if there is a guest cookie (= guest context) but the user is logged in, then the logged in should take priority. This is a function decision.

It implies technically to know whether the user is logged in, and apparently the only reliable way to check that is through `request.session.uid`.

`self.env.user` includes the public user so it's not enough, also it's not set with auth=None.

`self.env.user._is_public()` does not work for routes with auth=None (because it does ensure on on the user), which is the issue of this PR.

So instead `request.session.uid` is used.

But `request` does not work when called from a non-route (cron, upgrade, ...) so it has to be guarded too.

Reply:
Why would a route without auth call message_post as public (how do we do it ?), without setting explicitly author, email_from and guest. Request has nothing to do in message_post. It should receives parameters and acts accordingly. Caller has to give author and/or guest based on cookies, context or whatever he wants. Message_post should not be called like that, you should have a route ensuring it is done correctly or called with right parameters.

We should completely remove the reference to request in message_post, meaning cron / upgrade / ... are not broken anymore.

Reply:
About parameters: it takes kwargs and those are separated into message kwargs and notification kwargs. About everything that is a field is supported, so context manipulating parameters should not be necessary.